functionalization: avoid some unnecessary view_copy calls

bdhirsh opened the pull request:
When we're performing a sync on a `FuntionalTensorWrapper`, right now we first apply any updates to the base, and then we unconditionally regenerate the current tensor from the base.

This came up in the context of mobile using `functionalize()` and seeing an extra copy, so I have tests in a corresponding functorch PR: https://github.com/pytorch/functorch/pull/678/files#diff-449582f4f3e1ae76a73397c4b8ca62cc9dd5235dd831187f143180b498dee332R2872

cc @ZolotukhinM 

We can avoid some unnecessary copies going to the backend if we skip the regenerating when we know that there weren't any updates to apply.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527 [prototype] integrate functionalization <> LTC torchscript backend
* #75913 functionalization: introduce a "zero()" aten op
* **#75819 functionalization: avoid some unnecessary view_copy calls**
* #75717 fix unfold for meta tensors
* #75716 teach ivalue about List[Optional[Tensor]], fix fallbacks
* #75818 fix out= op handling for functionalization
* #75302 split out functionalization codegen to use view_copy operators

Differential Revision: [D35705380](https://our.internmc.facebook.com/intern/diff/D35705380)

bdhirsh committed:
avoid some unnecessary view_copy calls

[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: avoid some unnecessary view_copy calls"

When we're performing a sync on a `FuntionalTensorWrapper`, right now we first apply any updates to the base, and then we unconditionally regenerate the current tensor from the base.

This came up in the context of mobile using `functionalize()` and seeing an extra copy, so I have tests in a corresponding functorch PR: https://github.com/pytorch/functorch/pull/678/files#diff-449582f4f3e1ae76a73397c4b8ca62cc9dd5235dd831187f143180b498dee332R2872

cc @ZolotukhinM 

We can avoid some unnecessary copies going to the backend if we skip the regenerating when we know that there weren't any updates to apply.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: avoid some unnecessary view_copy calls"

When we're performing a sync on a `FuntionalTensorWrapper`, right now we first apply any updates to the base, and then we unconditionally regenerate the current tensor from the base.

This came up in the context of mobile using `functionalize()` and seeing an extra copy, so I have tests in a corresponding functorch PR: https://github.com/pytorch/functorch/pull/678/files#diff-449582f4f3e1ae76a73397c4b8ca62cc9dd5235dd831187f143180b498dee332R2872

cc @ZolotukhinM 

We can avoid some unnecessary copies going to the backend if we skip the regenerating when we know that there weren't any updates to apply.




[ghstack-poisoned]

ezyang reviewed (approved):
Nice! Having a test would be even better!

bdhirsh committed:
Update on "functionalization: avoid some unnecessary view_copy calls"

When we're performing a sync on a `FuntionalTensorWrapper`, right now we first apply any updates to the base, and then we unconditionally regenerate the current tensor from the base.

This came up in the context of mobile using `functionalize()` and seeing an extra copy, so I have tests in a corresponding functorch PR: https://github.com/pytorch/functorch/pull/678/files#diff-449582f4f3e1ae76a73397c4b8ca62cc9dd5235dd831187f143180b498dee332R2872

cc @ZolotukhinM 

We can avoid some unnecessary copies going to the backend if we skip the regenerating when we know that there weren't any updates to apply.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: avoid some unnecessary view_copy calls"

When we're performing a sync on a `FuntionalTensorWrapper`, right now we first apply any updates to the base, and then we unconditionally regenerate the current tensor from the base.

This came up in the context of mobile using `functionalize()` and seeing an extra copy, so I have tests in a corresponding functorch PR: https://github.com/pytorch/functorch/pull/678/files#diff-449582f4f3e1ae76a73397c4b8ca62cc9dd5235dd831187f143180b498dee332R2872

cc @ZolotukhinM 

We can avoid some unnecessary copies going to the backend if we skip the regenerating when we know that there weren't any updates to apply.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: avoid some unnecessary view_copy calls"

When we're performing a sync on a `FuntionalTensorWrapper`, right now we first apply any updates to the base, and then we unconditionally regenerate the current tensor from the base.

This came up in the context of mobile using `functionalize()` and seeing an extra copy, so I have tests in a corresponding functorch PR: https://github.com/pytorch/functorch/pull/678/files#diff-449582f4f3e1ae76a73397c4b8ca62cc9dd5235dd831187f143180b498dee332R2872

cc @ZolotukhinM 

We can avoid some unnecessary copies going to the backend if we skip the regenerating when we know that there weren't any updates to apply.




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35705380).

bdhirsh commented:
@pytorchmergebot merge this please
